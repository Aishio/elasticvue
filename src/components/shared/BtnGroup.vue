<template>
  <div :class="classes">
    <slot></slot>
  </div>
</template>

<script>
  export default {
    name: 'BtnGroup',
    props: {
      small: Boolean
    },
    computed: {
      classes () {
        return [
          'elevation-1',
          'btn-group',
          {'btn-group--small': this.small}
        ]
      }
    }
  }
</script>
