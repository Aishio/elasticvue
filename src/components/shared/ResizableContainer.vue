<template>
  <div ref="resizedWrapper" :style="style" class="resizable-container">
    <slot/>

    <div class="resizable-container__vertical-handler" @mousedown="onDragStart"/>
  </div>
</template>

<script>
  import BtnGroup from '@/components/shared/BtnGroup'

  export default {
    name: 'resizable-container',
    components: {
      BtnGroup
    },
    props: {
      initialWidth: {
        default: '100%',
        type: String
      },
      initialHeight: {
        default: 350,
        type: Number
      },
      stepSize: {
        default: 150,
        type: Number
      }
    },
    data () {
      return {
        width: this.initialWidth,
        height: this.initialHeight,
        dragStartY: 0,
        dragStartHeight: this.initialHeight,
        resizing: false
      }
    },
    computed: {
      style () {
        return `width: ${this.width}; height: ${this.height}px`
      }
    },
    methods: {
      triggerResize () {
        if (typeof window === 'undefined') return
        window.dispatchEvent(new Event('resize'))
      },
      onDragStart (e) {
        window.addEventListener('mouseup', this.onDragEnd)
        window.addEventListener('mousemove', this.onDrag)
        this.resizing = true
        this.dragStartY = e.pageY
        this.dragStartHeight = this.height
      },
      onDrag (e) {
        if (this.resizing) {
          const newHeight = this.dragStartHeight + e.pageY - this.dragStartY
          if (newHeight > this.initialHeight) {
            this.height = newHeight
            const distanceToTop = e.clientY
            const distanceToBottom = window.innerHeight - e.clientY
            const scrollOffset = 100
            if (distanceToBottom < scrollOffset) {
              // scroll down automatically
              window.scrollBy(0, scrollOffset - distanceToBottom)
            } else if (distanceToTop < 100) {
              // scroll up automatically
              window.scrollBy(0, -(scrollOffset - distanceToTop))
            }
          }
        }
      },
      onDragEnd () {
        this.resizing = false
        this.dragStartY = 0
        this.$nextTick(() => {
          if (typeof window !== 'undefined') {
            this.triggerResize()
          }
        })
        window.removeEventListener('mouseup', this.onDragEnd)
        window.removeEventListener('mousemove', this.onDrag)
      }
    }
  }
</script>
