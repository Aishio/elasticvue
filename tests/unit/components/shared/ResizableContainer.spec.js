import Vuetify from 'vuetify'
import VBtn from 'vuetify/es5/components/VBtn'
import VIcon from 'vuetify/es5/components/VIcon'
import VGrid from 'vuetify/es5/components/VGrid'
import ResizableContainer from '@/components/shared/ResizableContainer'
import BtnGroup from '@/components/shared/BtnGroup'
import { createLocalVue, mount } from '@vue/test-utils'
import '../../mocks/codeEditorMocks'

describe('components/shared/ResizableContainer.vue', () => {
  let localVue

  beforeEach(() => {
    localVue = createLocalVue()
    localVue.use(Vuetify, {
      components: {
        BtnGroup,
        VBtn,
        VIcon,
        VGrid
      }
    })
  })

  it('should render correct default contents', () => {
    const wrapper = mount(ResizableContainer, { localVue })
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should render correct initialHeight', () => {
    const wrapper = mount(ResizableContainer, {
      localVue: localVue,
      propsData: {
        initialHeight: 500
      }
    })
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should increase the height on drag', () => {
    const wrapper = mount(ResizableContainer, {
      localVue: localVue,
      attachToDocument: true
    })
    const handler = wrapper.find('div.resizable-container__vertical-handler')

    const startY = 0
    const endY = 100

    handler.trigger('mousedown', { pageY: startY })
    handler.trigger('mousemove', { pageY: endY })
    handler.trigger('mouseup')
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should increase the height on drag and scroll if needed', () => {
    const wrapper = mount(ResizableContainer, {
      localVue: localVue,
      attachToDocument: true
    })
    const handler = wrapper.find('div.resizable-container__vertical-handler')

    const startY = 0
    const endY = 1000

    handler.trigger('mousedown', { pageY: startY, clientY: startY })
    handler.trigger('mousemove', { pageY: endY, clientY: endY })
    handler.trigger('mouseup')
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should not decrease the height lower then intialHeight', () => {
    const wrapper = mount(ResizableContainer, {
      localVue: localVue,
      attachToDocument: true
    })
    const handler = wrapper.find('div.resizable-container__vertical-handler')

    const startY = 100
    const endY = 0

    handler.trigger('mousedown', { pageY: startY })
    handler.trigger('mousemove', { pageY: endY })
    handler.trigger('mouseup')
    expect(wrapper.element).toMatchSnapshot()
  })
})
