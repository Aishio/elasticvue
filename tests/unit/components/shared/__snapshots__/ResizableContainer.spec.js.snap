// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/shared/ResizableContainer.vue should increase the height on drag 1`] = `
<div
  class="resizable-container"
  style="width: 100%; height: 450px;"
>
   
  <div
    class="resizable-container__vertical-handler"
  />
</div>
`;

exports[`components/shared/ResizableContainer.vue should increase the height on drag and scroll if needed 1`] = `
<div
  class="resizable-container"
  style="width: 100%; height: 1350px;"
>
   
  <div
    class="resizable-container__vertical-handler"
  />
</div>
`;

exports[`components/shared/ResizableContainer.vue should not decrease the height lower then intialHeight 1`] = `
<div
  class="resizable-container"
  style="width: 100%; height: 350px;"
>
   
  <div
    class="resizable-container__vertical-handler"
  />
</div>
`;

exports[`components/shared/ResizableContainer.vue should render correct default contents 1`] = `
<div
  class="resizable-container"
  style="width: 100%; height: 350px;"
>
   
  <div
    class="resizable-container__vertical-handler"
  />
</div>
`;

exports[`components/shared/ResizableContainer.vue should render correct initialHeight 1`] = `
<div
  class="resizable-container"
  style="width: 100%; height: 500px;"
>
   
  <div
    class="resizable-container__vertical-handler"
  />
</div>
`;
